File: apps/frontend/lib/models/record.dart
import 'package:uuid/uuid.dart';

class HealthRecord {
  final int? localId;
  final String clientMsgId;
  final String userId;
  final String recordType;
  final double recordValue; // 热量(kcal) 或 体重(kg) 或 睡眠(hours)
  final double? duration; // 运动时长（分钟）
  final String unit;
  final String recordDate;
  final String? notes;
  final int syncStatus;
  final String? exerciseName; // 🚀 新增：运动名称

  // 🚀 【新增】：精细化饮食打卡 5 大字段
  final String? mealType;
  final String? foodName;
  final double? carbsG;
  final double? proteinG;
  final double? fatG;

  HealthRecord({
    this.localId,
    String? clientMsgId,
    required this.userId,
    required this.recordType,
    required this.recordValue,
    this.duration,
    required this.unit,
    required this.recordDate,
    this.notes,
    this.syncStatus = 0,
    // 👉 饮食专属字段
    this.mealType,
    this.foodName,
    this.carbsG,
    this.proteinG,
    this.fatG,
    this.exerciseName,
  }) : clientMsgId = clientMsgId ?? const Uuid().v4();

  factory HealthRecord.fromMap(Map<String, dynamic> map) {
    return HealthRecord(
      localId: map['local_id'] as int?,
      clientMsgId: map['client_msg_id'] as String,
      userId: map['user_id'] as String,
      recordType: map['record_type'] as String,
      recordValue: (map['record_value'] as num).toDouble(),
      duration: map['duration'] != null
          ? (map['duration'] as num).toDouble()
          : null,
      unit: map['unit'] as String,
      recordDate: map['record_date'] as String,
      notes: map['notes'] as String?,
      syncStatus: map['sync_status'] as int? ?? 0,
      exerciseName: map['exercise_name'] as String?,

      // 🚀 精细化饮食字段（严格驼峰映射与 num 安全强转）
      mealType: map['meal_type'] as String?,
      foodName: map['food_name'] as String?,
      carbsG: map['carbs_g'] != null
          ? (map['carbs_g'] as num).toDouble()
          : null,
      proteinG: map['protein_g'] != null
          ? (map['protein_g'] as num).toDouble()
          : null,
      fatG: map['fat_g'] != null ? (map['fat_g'] as num).toDouble() : null,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'local_id': localId,
      'client_msg_id': clientMsgId,
      'user_id': userId,
      'record_type': recordType,
      'record_value': recordValue,
      'duration': duration,
      'unit': unit,
      'record_date': recordDate,
      'notes': notes,
      'sync_status': syncStatus,

      // 🚀 写入 SQLite (严格下划线)
      'meal_type': mealType,
      'food_name': foodName,
      'carbs_g': carbsG,
      'protein_g': proteinG,
      'fat_g': fatG,
      'exercise_name': exerciseName,
    };
  }

  Map<String, dynamic> toJson() {
    return {
      'client_msg_id': clientMsgId,
      'user_id': userId,
      'record_type': recordType,
      'record_value': recordValue,
      'duration': duration,
      'unit': unit,
      'record_date': recordDate,
      'notes': notes,

      // 🚀 发送给云端 (严格 JSON 契约)
      'meal_type': mealType,
      'food_name': foodName,
      'carbs_g': carbsG,
      'protein_g': proteinG,
      'fat_g': fatG,
      'exercise_name': exerciseName,
    };
  }

  HealthRecord copyWith({
    int? localId,
    String? clientMsgId,
    String? userId,
    String? recordType,
    double? recordValue,
    double? duration,
    String? unit,
    String? recordDate,
    String? notes,
    int? syncStatus,
    String? mealType,
    String? foodName,
    double? carbsG,
    double? proteinG,
    double? fatG,
    String? exerciseName,
  }) {
    return HealthRecord(
      localId: localId ?? this.localId,
      clientMsgId: clientMsgId ?? this.clientMsgId,
      userId: userId ?? this.userId,
      recordType: recordType ?? this.recordType,
      recordValue: recordValue ?? this.recordValue,
      duration: duration ?? this.duration,
      unit: unit ?? this.unit,
      recordDate: recordDate ?? this.recordDate,
      notes: notes ?? this.notes,
      syncStatus: syncStatus ?? this.syncStatus,
      mealType: mealType ?? this.mealType,
      foodName: foodName ?? this.foodName,
      carbsG: carbsG ?? this.carbsG,
      proteinG: proteinG ?? this.proteinG,
      fatG: fatG ?? this.fatG,
      exerciseName: exerciseName ?? this.exerciseName,
    );
  }
}

File: apps/backend/app/models/record.py
"""
app/models/record.py

核心数据流向：映射关系型数据表 health_records 到 Python 实例操作。
核心逻辑：摒弃自增采用前端生成的物理主键 client_msg_id，提供高度灵活但 nullable=True 防破坏的细粒度打卡字段。
具体职责：大体量流水登记核心表，包容饮食、睡眠、运动全维度的行为审计记录，建立针对性复合索引支持高效按日聚合。
"""
from sqlalchemy import Column, String, Numeric, TIMESTAMP, text, Index
from app.db.base import Base


class HealthRecord(Base):
    """
    健康打卡记录表（登记员模式）：
    不再使用自增 ID，直接以前端生成的 UUID (client_msg_id) 作为物理主键。
    """
    __tablename__ = "health_records"

    # 🌟 物理主键：前端 UUID
    client_msg_id = Column(
        String(36),
        primary_key=True,
        nullable=False,
        comment="前端生成的唯一UUID，既是业务ID也是物理主键"
    )

    # 租户隔离
    user_id = Column(String(64), nullable=False, index=True, comment="关联用户中心的主键")

    # 🌟 业务字段更新：加入 WEIGHT 类型说明
    record_type = Column(
        String(20),
        nullable=False,
        comment="类型: DIET(饮食), EXERCISE(运动), SLEEP(睡眠), WEIGHT(体重)"
    )

    # 🌟 数值精度：(10, 2) 足够覆盖 0.01 到 99,999,999.99，非常稳
    record_value = Column(Numeric(10, 2), nullable=False, comment="数值：热量(kcal)、时长(小时)或体重(kg)")

    # 🌟 单位说明更新：加入 kg
    unit = Column(String(10), nullable=False, comment="单位：如 kcal, hours, kg")
    duration = Column(Numeric(10, 2), nullable=True, comment="补充时长(数值)，如运动分钟数")
    # 🌟 新增：精细化饮食打卡字段 (向下兼容，全设为 nullable=True)
    meal_type = Column(String(20), nullable=True, comment="餐别: BREAKFAST, LUNCH, DINNER, SNACK")
    food_name = Column(String(100), nullable=True, comment="食物名称")
    carbs_g = Column(Numeric(6, 2), nullable=True, comment="碳水化合物(g)")
    protein_g = Column(Numeric(6, 2), nullable=True, comment="蛋白质(g)")
    fat_g = Column(Numeric(6, 2), nullable=True, comment="脂肪(g)")
    # 🆕 运动名称（仅 EXERCISE 类型使用）
    exercise_name = Column(String(100), nullable=True, comment="运动名称（仅EXERCISE类型使用）")
    notes = Column(String(255), nullable=True, comment="用户附加备注")

    # 时间审计
    record_date = Column(TIMESTAMP, nullable=False, comment="实际打卡发生的业务时间")
    created_at = Column(
        TIMESTAMP,
        server_default=text("CURRENT_TIMESTAMP"),
        comment="记录入库的审计时间"
    )

    # 🚀 复合索引优化
    __table_args__ = (
        Index("idx_user_date_type", "user_id", "record_date", "record_type"),
    )